scripts/exp/extract-deps.helpers.ts

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as tmp from 'tmp';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as rollup from 'rollup';
+import * as terser from 'terser';
+import * as ts from 'typescript';
+
+/** Contains the dependencies and the size of their code for a single export. */
+export type ExportData = { dependencies: string[]; sizeInBytes: number };
+
+/**
+ * This functions builds a simple JS app that only depends on the provided
+ * export. It then uses Rollup to gather all top-level classes and functions
+ * that that the export depends on.
+ *
+ * @param exportName The name of the export to verify
+ * @param jsBundle The file name of the source bundle that contains the export
+ * @return A list of dependencies for the given export
+ */
+export async function extractDependencies(
+  exportName: string,
+  jsBundle: string
+): Promise<string[]> {
+  const { dependencies } = await extractDependenciesAndSize(
+    exportName,
+    jsBundle
+  );
+  return dependencies;
+}
+
+/**
+ * Helper for extractDependencies that extracts the dependencies and the size
+ * of the minified build.
+ */
+export async function extractDependenciesAndSize(
+  exportName: string,
+  jsBundle: string
+): Promise<ExportData> {
+  const input = tmp.fileSync().name + '.js';
+  const output = tmp.fileSync().name + '.js';
+
+  // JavaScript content that exports a single API from the bundle
+  const beforeContent = `export { ${exportName} } from '${path.resolve(
+    jsBundle
+  )}';`;
+  fs.writeFileSync(input, beforeContent);
+
+  // Run Rollup on the JavaScript above to produce a tree-shaken build
+  const bundle = await rollup.rollup({
+    input,
+    external: id => id.startsWith('@firebase/')
+  });
+  await bundle.write({ file: output, format: 'es' });
+
+  const dependencies = extractFunctionsAndClasses(output);
+  dependencies.sort();
+
+  // Extract size of minified build
+  const afterContent = fs.readFileSync(output, 'utf-8');
+  const { code } = terser.minify(afterContent, {
+    output: {
+      comments: false
+    },
+    mangle: false,
+    compress: false
+  });
+
+  fs.unlinkSync(input);
+  fs.unlinkSync(output);
+
+  return { dependencies, sizeInBytes: Buffer.byteLength(code!, 'utf-8') };
+}
+
+/**
+ * Extracts all top-level function and classes using the TypeScript compiler
+ * API.
+ */
+export function extractFunctionsAndClasses(jsFile: string): string[] {
+  const program = ts.createProgram([jsFile], { allowJs: true });
+  const sourceFile = program.getSourceFile(jsFile);
+  if (!sourceFile) {
+    throw new Error('Failed to parse file: ' + jsFile);
+  }
+
+  const exports: string[] = [];
+  ts.forEachChild(sourceFile, node => {
+    if (ts.isFunctionDeclaration(node)) {
+      exports.push(node.name!.text);
+    } else if (ts.isClassDeclaration(node)) {
+      exports.push(node.name!.text);
+    }
```

Comment:
It'd be interesting to see variable declarations too, e.g. the error message map.
It'd be nice to keep the types of the declarations and show them in the report, but it can be done in a separate PR.